src/upkeeps/UpKeepManager.sol

```diff
@@ -0,0 +1,496 @@
+
+        uint256 linkRegistryBal = LINK.balanceOf(address(this));
+        if (linkRegistryBal < linkAmount) {
+            _swapEthForLink(linkAmount - linkRegistryBal);
+        }
+
+        /// @dev checks CL registry state before registering
+        (IKeeperRegistry.State memory state,,) = CL_REGISTRY.getState();
+        uint256 oldNonce = state.nonce;
+
+        bytes memory data = abi.encodeCall(
+            IKeeperRegistrar.register,
+            (
+                _name,
+                bytes(""),
+                _targetAddress,
+                uint32(_gasLimit),
+                address(this),
+                bytes(""),
+                uint96(linkAmount),
+                0,
+                address(this)
+            )
+        );
+
+        LINK.transferAndCall(KEEPER_REGISTRAR, linkAmount, data);
+
+        (state,,) = CL_REGISTRY.getState();
+        uint256 newNonce = state.nonce;
+
+        if (newNonce == oldNonce + 1) {
+            upkeepID_ = uint256(
+                keccak256(abi.encodePacked(blockhash(block.number - 1), address(CL_REGISTRY), uint32(oldNonce)))
+            );
+        } else {
+            revert NotAutoApproveKeeper();
+        }
+    }
+
+    ////////////////////////////////////////////////////////////////////////////
+    // PUBLIC VIEW
+    ////////////////////////////////////////////////////////////////////////////
+
+    /// @notice Checks whether an upkeep is to be performed.
+    /// @dev The calldata is encoded with the targetted member address
+    /// @return upkeepNeeded_ A boolean indicating whether an upkeep is to be performed.
+    /// @return performData_ The calldata to be passed to the upkeep function.
+    function checkUpkeep(bytes calldata)
+        external
+        view
+        override
+        whenNotPaused
+        returns (bool upkeepNeeded_, bytes memory performData_)
+    {
+        address[] memory members = getMembers();
+        bool underFunded;
+
+        uint256 membersLength = members.length;
+        if (membersLength > 0) {
+            // NOTE: loop thru members to see which is underfunded
+            for (uint256 i; i < membersLength;) {
+                uint256 upKeepId = membersInfo[members[i]].upKeepId;
+                (, underFunded) = _isUpKeepIdUnderFunded(upKeepId);
+                if (underFunded) {
+                    (,,,,,, uint64 maxValidBlocknumber,) = CL_REGISTRY.getUpkeep(upKeepId);
+                    // NOTE: helps filtering those which `cancelUpkeep` has being initiated
+                    if (maxValidBlocknumber == UINT64_MAX) {
+                        upkeepNeeded_ = true;
+                        performData_ = abi.encode(members[i]);
+                        break;
+                    }
+                }
+                unchecked {
+                    ++i;
+                }
+            }
+        }
+
+        // NOTE: to avoid overwritten an `upKeep` meant for registration, check boolean
+        if (!upkeepNeeded_) {
```

Comment:
it is a good point, how i envisioned is that `minRoundsTopUp` variable together with each particular `upKeepId` based on their gas needs to find very unlikely that even two members could be under-funded in the same block and blocking each other, also a block diff, should not be a deal-breaker for these keeper tasks so far. For liquidations or alike would be something to dig deeper to optimised if it was the case.

but it is a concern, which let me think that given the `UpKeepManager` to have a greater importance in the ladder against the members, since that one getting under-funded will make the whole process stuck, sounds worth to prio it in the order of the `checkUpKeep` just to have peace of mind

Reply:
see 75b3b2897f4de641a98970066b20dc9f177441e6

Reply:
> see [75b3b28](https://github.com/Badger-Finance/badger-avatars/commit/75b3b2897f4de641a98970066b20dc9f177441e6)

do we need to early return right after if `monitoringUpKeepId` is under-funded? https://github.com/Badger-Finance/badger-avatars/blob/75b3b2897f4de641a98970066b20dc9f177441e6/src/upkeeps/UpKeepManager.sol#L465-L468

otherwise the `performData_` may get overwritten by under-funded member upKeepId

Reply:
added early explicit return in ccb918e9abf852c4b90b560c5a88f93ce49fea66